proposals/0000-React-Native-Test-App-as-a-Package.md

```diff
@@ -0,0 +1,285 @@
+  development and CI agents
+- Reduce or eliminate the friction to upgrade React Native, and try out release
+  candidates
+- Reduce or eliminate the overhead of adding support for new platforms
+
+With common test apps, we will also be able to provide additional value for
+everyone simultanously. For instance:
+
+- We will handle upgrading of the test apps to support the latest mobile SDK
+- We can provide a common infrastructure for native testing
+- We can provide optional modules for commonly used services, such as auth or
+  App Center integration
+
+## Detailed design
+
+Given an existing React Native app/library, we imagine the ideal flow for
+**initial integration** with the native test apps should be:
+
+1. Create folder/package for the test app to live in.
+   - This package should have a dependency on the app/library to be tested, and
+     on the test app package.
+2. Create a manifest to declare entry points. Optionally, it should also be
+   possible to declare initial props, native source files and resources.
+   - This is a **one-time cost**, and should be stable across versions.
+   - If there is a single entry point, the app may boot directly into this app.
+   - Otherwise, if there are multiple entry points, we will show them in a list.
+     - It should still be possible to have the app boot into one of these via
+       deep linking.
+   - Native dependencies should be taken care of by auto-linking behind the
+     scenes.
+3. Run `npm install` inside this folder as one would usually do.
+4. Depending on the desired target platforms, the next steps are:
+   - Android:
+     - Make minimal changes to `build.gradle` (one-time cost)
+     - Open the project in Android Studio
+   - iOS:
+     - Make minimal changes to `Podfile` (one-time cost)
+     - Run `pod install`
+     - Open the newly generated Xcode workspace
+   - Windows:
+     - <@afoxman or @acoates-ms to fill in>
+   - macOS:
+     - See steps for iOS
+
+A hypothetical package should end up looking something like below:
+
+```
+my-awesome-component
+├── example
+│   ├── App.js
+│   ├── build.gradle (new/modified)
+│   ├── package.json
+│   ├── Podfile (new/modified)
+│   └── yarn.lock
+├── my-awesome-component.podspec
+├── my-awesome-component-dev.podspec (new, optional)
+├── package.json
+├── src
+└── yarn.lock
+```
+
+For a hypothetical monorepo:
+
+```
+my-monorepo
+├── package.json
+├── packages
+│   ├── my-awesome-component-android
+│   │   ├── build.gradle (new/modified)
+│   │   ├── package.json
+│   │   └── src
+│   ├── my-awesome-component-core
+│   │   ├── package.json
+│   │   └── ...
+│   └── my-awesome-component-ios
+│       ├── my-awesome-component.podspec
+│       ├── my-awesome-component-dev.podspec (new, optional)
+│       ├── package.json
+│       ├── Podfile (new/modified)
+│       └── src
+└── yarn.lock
+```
+
+Platform specific project files should no longer be checked in. They are either
```

Comment:
There were a lot of discussions about "generating native files at a build time" in the past, unfortunately, all of them ended up being stuck after a while.

You may find this issue https://github.com/facebook/react-native/issues/6292 useful for some additional context.